symengine/series_generic.cpp

```diff
@@ -1,166 +1,197 @@
-        var_{var}, prec_{precision} {
+        SeriesBase(UnivariateExprPolynomial(univariate_polynomial(var, max, convert_map(dict))), var->get_name(), precision) {}
+
+UnivariateSeries::UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const map_uint_mpz& dict) :
+        SeriesBase(convert_poly(std::move(dict)), var->get_name(), precision) {}
+
+UnivariateSeries::UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const std::vector<mpz_class> &v) :
+        SeriesBase(convert_vector(v), var->get_name(), precision) {}
 
-    poly_ = univariate_polynomial(var_, max, std::move(dict));
+RCP<const UnivariateSeries> UnivariateSeries::series(const RCP<const Basic> &t, const std::string &x, unsigned int prec) {
+    SeriesVisitor<UnivariateExprPolynomial, Expression, UnivariateSeries> visitor(UnivariateExprPolynomial(std::stoi(x)), x, prec);
+    return visitor.series(t);
 }
 
-UnivariateSeries::UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const map_uint_mpz& dict) :
-        var_{var}, prec_{precision} {
+std::size_t UnivariateSeries::__hash__() const {
+    return p_.__hash__() + std::size_t(prec_ * 84728863L);
+}
 
-    map_uint_mpz dict_trunc;
-    unsigned int max = 0;
-    std::copy_if(dict.begin(), dict.end(), std::inserter(dict_trunc, dict_trunc.end()),
-        [&](const map_uint_mpz::value_type i)
-        {
-            if (i.first < prec_) {
-                if (max < i.first)
-                    max = i.first;
-                return true;
-            }
-            return false;
-        } );
-    poly_ = univariate_polynomial(var_, max, std::move(dict_trunc));
+int UnivariateSeries::compare(const Basic &other) const {
+    if (not is_a<UnivariateSeries>(other))
+        throw std::domain_error("cannot compare with UnivariateSeries");
+    const UnivariateSeries &o = static_cast<const UnivariateSeries &>(other);
+    return p_ == (o.p_);
 }
 
-UnivariateSeries::UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const std::vector<mpz_class> &v) :
-        var_{var}, prec_{precision} {
+RCP<const Basic> UnivariateSeries::as_basic() const {
+    return p_.get_basic();
+}
 
-    std::vector<mpz_class> vtrunc;
-    std::copy_if(v.begin(), v.end(), std::back_inserter(vtrunc),
-        [&](decltype(v[0]) i) { return i < prec_; } );
-    poly_ = UnivariatePolynomial::create(var_, vtrunc);
+umap_int_basic UnivariateSeries::as_dict() const {
+    umap_int_basic map;
+    for (int i = 0; i <= degree_; i++) 
+       map[i] = p_.get_univariate_poly()->dict_.at(i).get_basic();
+    return map;
 }
 
-bool UnivariateSeries::is_canonical(const UnivariatePolynomial& poly, const unsigned int &prec) const
-{
-    return true;
+RCP<const Basic> UnivariateSeries::get_coeff(int deg) const {
+    return p_.get_univariate_poly()->get_dict().at(deg).get_basic();
```

Comment:
If `deg` is not there, this will throw an exception. When `deg` is not there, we need to return zero.